PC Relate with Persist annotations [PCRI-3]

danking opened the pull request:
The third in a series of PCRelate Improvements. Depends on #2249 

Sprinkling `cache` on any RDD which is used more than once dramatically improved runtime. On a benchmark program (included below) PCRelate took 40 seconds with four cores on 1000 samples and 10,000 variants.

```
from hail import *
from timeit import default_timer as timer

hc = HailContext()
vds = hc.balding_nichols_model(20, 1000, 10000).repartition(10).persist()
vds.count()

start = timer()
vds.pc_relate(5, 0.01, min_kinship=0.1, desire="all").count()
end = timer()
print(end-start)
```

jigold commented on src/main/scala/is/hail/methods/PCRelate.scala:
```diff
@@ -243,16 +247,16 @@ class PCRelate(maf: Double, blockSize: Int) extends Serializable {
+    val homref = (dm.map2 { (g, mu) =>
       if (badgt(g) || badmu(mu) || g != 0.0) 0.0 else 1.0
-    } (g, mu)
+    } (g, mu)).cache
```
choose one of `cache` versus `cache()`.

jigold commented on src/main/scala/is/hail/methods/PCRelate.scala:
```diff
@@ -190,12 +190,16 @@ class PCRelate(maf: Double, blockSize: Int) extends Serializable {
     mu <= maf || mu >= antimaf || mu <= 0.0 || mu >= 1.0
   def badgt(gt: Double): Boolean =
     gt != 0.0 && gt != 1.0 && gt != 2.0
+  private def gram(m: M): M = {
```
I prefer spaces between method definitions. Your call.

jigold commented on src/main/scala/is/hail/methods/PCRelate.scala:
```diff
@@ -168,28 +190,39 @@ class PCRelate(maf: Double, blockSize: Int) extends Serializable {
-          0.0
-        else
-          mu * (1.0 - mu)
+    val variance = dm.map2 { (g, mu) =>
```
Should the `variance` be cached?

danking commented:
I've addressed the comments, but I can't run the benchmark anymore, I get weird errors about partitions being empty that I expected to be non-empty. I will continue to investigate.

danking commented:
Benchmark is now like 80s. Not sure why it got slower from before.

danking commented:
I'll re-open when #2270 lands because this builds on that.
